VK tests, GitHub actions, GitHub pages, allure report and linters

Tony763 opened the pull request:
Hi,
 I finally managed to finish next step in dividing and applying changes from PR #38.
Check [README.md](tests/README.md) for some detailed information
Actual report can be seen [there](https://tony763.github.io/skill-homeassistant/)

#### Description
Fix old unittests
Remove old tests
Add VK tests (behave)
Add GitHub Actions workflow: Contain linters, VK tests, allure report
Add docstrings
Fix various errors found by linters
Add `Readme.md` to test directory
Add .venv and .vscode to .gitignore

#### Type of PR
- [x] Bugfix
- [x] Feature implementation
- [x] Refactor of code (without functional changes)
- [x] Documentation improvements
- [x] Test improvements

#### Note
PR contain partially changes from #74 to fix pylint errors, but not all. It is still valid PR.
Changes are divided by commits.

Tony763 committed:
Fix old unittests

Tony763 committed:
Remove old tests

Tony763 committed:
Add behave tests

Tony763 committed:
Add github actions workflow with HA/Mycroft settings

Tony763 committed:
Add Readme for new tests

Tony763 committed:
Fix pycodestyle errors

Tony763 committed:
Fix flake8 errors

Tony763 committed:
Fix isort errors

krisgesling commented on __init__.py:
```diff
@@ -579,12 +598,14 @@ def handle_fallback(self, message):
+        super().shutdown()
 
     def stop(self):
-        pass
+        """Just pass on stop"""
```
Can remove this entirely. Its only needed if the Skill actually needs to respond to "stop" like commands.

Tony763 replied:
Will remove. Commit prepared.

Tony763 replied:
Done

krisgesling commented on ha_client.py:
```diff
@@ -1,26 +1,35 @@
+        port_number = config['port_number']
         if self.ssl:
-            self.url = "https://{}".format(host)
+            self.url = "https://{}".format(ip_address)
```
I saw in `__init__.py` you started using f-strings. Now that we're Python3.6+ we can use those so I wonder if we want to do that consistently throughout?

Tony763 replied:
Definitely yes. I overlooked these. I also changed them in logging so  f -Strings are only used. Commit prepared.

Tony763 replied:
Actually have to switch all logging back to `%` as logging lib is optimized for that usage and linters complained.

Tony763 replied:
Done

krisgesling commented on ha_client.py:
```diff
@@ -92,6 +102,7 @@ def find_entity(self, entity, types):
                 except KeyError:
                     pass
             return best_entity
+        return None
```
Given we instantiate `best_entity` as `None` before the loop, is it worth just returning `best_entity` in all cases? 

Essentially remove one indent level from line 104

Tony763 replied:
True, changed. Commit prepared.

Tony763 replied:
Done

krisgesling commented on test/behave/07_set_brightness.feature:
```diff
@@ -0,0 +1,5 @@
+Feature: set brightness
```
One option is to group these tests within fewer feature files, eg you can have all the brightness setting scenario's in a single `set_brightness.feature` file.

Though this was also useful to see that all the prior integration tests have been moved across.

Tony763 replied:
Grouping is one way, but I would rather keep one file per intent and when we find conflict with other intent, add tests for it to them.

For exampel: Issue #64 conflict with `Where is` skill. I have prepared fix for it and also I will add test to check that this skill won't interfere with HA skill and HA skill will not interfere with Where is skill.

With this approach, you have clean and readable tests and you see at first glance, which intents had conflicts as they have more scenarios.

krisgesling commented on test/behave/02_turn_on_switch.feature:
```diff
@@ -0,0 +1,5 @@
+Feature: turn on
+  Scenario: turn on switch
+    Given an English speaking user
+    When the user says "can you turn on Mycroft sw please"
+	  Then "homeassistant" should reply with dialog from "homeassistant.device.on.dialog"
```
With behave tests, we want to set any expectations of the systems state using a custom `Given` step. 

You can see an example from the Timer Skill here:
https://github.com/MycroftAI/mycroft-timer/blob/38ccb2c362fe5ecc228797ef4102d0a62f7107fd/test/behave/steps/timer.py#L21

This means we aren't relying on tests to run in any particular order, and that if one test fails, it doesn't cause the following tests to fail where they might have passed.

Tony763 replied:
Understood, that is reasonable. I will try to add function to set entity to default state so if one test fail, other will still run. Otherwise I will have to create separate entity for each test, doable, but not so convenient.

Tony763 replied:
As I want to also check communication and handling by HA side, I cannot create any function to set them back. But I created separate entity for each test, so the can run on their own and do not interact with each other.

krisgesling replied:
That sounds like a good solution

krisgesling commented on unittests/test_haclient.py:
```diff
@@ -1,145 +0,0 @@
-from unittest import TestCase
```
Are you sure about removing the unit tests? 

I see them performing a different purpose than the behave/Voight Kampff integration tests. 

Unit tests should tell us that specific functions are performing as expected. That is given the same input, they return the expected output, or make the expected calls to some other function.

The integration tests on the other hand, tell us that with the entire system in place, if a user speaks a certain utterance, then the overall behaviour of the system is what we expect. We can assume that if the overall behaviour is correct that everything underneath must also be correct but that's not always the case. If something does go wrong, often unit tests are much more granular and help us to pin point where that problem exists.

Anyway, I'm not the one maintaining the unit tests to match new functionality in the Skill so I don't want to dictate how testing is done, just wanted to raise the question and make sure the unit tests don't have value still?

Tony763 replied:
True. I will keep them and fix them.

krisgesling commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,243 @@
+    - uses: actions/checkout@v2
+
+    # Setup python
+    - name: Set up Python 3.7
+      uses: actions/setup-python@v2
+      with:
+        python-version: 3.7
```
Do we want to support 3.6-3.9?

Tony763 replied:
It is possible, but it run then in parallel. It's fast, but it makes conflict with uploading reports/logs to gh-pages. I will try this but probably in a new PR.

krisgesling commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,243 @@
+        chmod +x /opt/mycroft-core/dev_setup.sh
+        /opt/mycroft-core/dev_setup.sh --allow-root -ci -sm -p ${{ env.PATH_TO_PYTHON }}
+        source $MYCROFT_VENV
+        /opt/mycroft-core/bin/mycroft-pip install -r ${{ github.workspace }}/requirements.txt
```
`dev_setup.sh` should install all Mycroft-core's dependencies.

Is this line installing the Skills requirements?

Tony763 replied:
This install skill requirements. `${{ github.workspace }}` point to root of skill.

krisgesling commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,243 @@
+      run: |
+        source $MYCROFT_VENV
+        if [ -f requirements.txt ]; then pip install -r requirements.txt; fi
+        pip install pylint
```
Or is this the Skill's requirements.txt?

Maybe we can improve the name of this step?

Tony763 replied:
It installed skill requirement second time. Removed. Now it only install some pip packages needed by workflow.

krisgesling commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,243 @@
+
+    # Wait for MC to start
+    - name: Mycroft - Wait for start
+      run: sleep 4m
```
Would be good to listen on the bus for the final `mycroft.ready` message, but sleeping for a while is fine for now.

krisgesling replied:
Actually this is possibly not needed anyway - the VK test runner will automatically wait for the system to be ready before it starts the tests.

Tony763 replied:
With Travis, I had problem that HA was not ready at time I started tests. I changed order of workflow and it seem that this problem is gone. Removed.

krisgesling commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,243 @@
+    # Open MC virtual enviroment
+    - name: MC - open virtual env
+      run: source /opt/mycroft-core/.venv/bin/activate
+
```
Does this add any pairing info? 
It shouldn't need any Selene services to operate, I'm just wondering if the Pairing Skill prompting to pair might cause issues with the tests?

krisgesling replied:
Seems to be passing fine on your fork so going to assume this isn't an issue until it becomes one. We could also blacklist the Pairing Skill.

Tony763 replied:
I just remove it.
```
- name: Mycroft - Remove pairing skill
      run: /opt/mycroft-core/bin/mycroft-msm remove mycroft-pairing
      continue-on-error: true
```

krisgesling replied:
This should work as it's getting installed and run all within 1 hour. But it would get reinstalled on the next update cycle of MSM if it's not blacklisted. Probably will be fine for this purpose but worth keeping in mind if we see longer runs or strange behaviour.

Tony763 replied:
Oh, I see. If MSM update, pairing skill is installed back. I tried to black list it with `mycroft-config set skills.blacklisted_skills ["mycroft-pairing"]`, but I am not sure if it disabled it correctly. Will have to test it on other branch.

krisgesling commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,243 @@
+    - name: MC - open virtual env
+      run: source /opt/mycroft-core/.venv/bin/activate
+
+    #######################################################################################
```
If we set the mycroft config tts.module to dummy the tests will all run faster as it doesn't try to generate audio for each spoken dialog.

Could either provide a system/user level `mycroft.conf` in the test files, or set the value in one of these steps:
```
run: /opt/mycroft-core/bin/mycroft-config set tts.module dummy
```

Tony763 replied:
Perfect, adding `mycroft-config set` is easiest for now, if we will need to add more settings, `mycroft.conf` can be added later. Commit prepared.

Tony763 replied:
Done

krisgesling commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,243 @@
+    # Job status - publish pylint
+    - name: State - pylint
+      if: always()
+      uses: Sibz/github-status-action@v1
```
Where does this get published?

Tony763 replied:
State of linter run is published as status to PR or commit. Details bellow.

![image](https://user-images.githubusercontent.com/5598750/135929602-274f2b1d-83a9-47e9-8b18-055adb106f46.png)

krisgesling replied:
Ah right - because they're all in the same job they would get reported as a single success / failure. This treats them as a separate job but you only have to setup mycroft-core once.

Tony763 replied:
Exactly, I am preparing also change that will add 'Detail' to each linter status and bring You to exact place in workflow log.

Tony763 replied:
Done, `Details` contain direct link to log section with selected linter.

![image](https://user-images.githubusercontent.com/5598750/136453299-389f8cd5-b895-467a-b814-2d39b17af89e.png)
`https://github.com/Tony763/skill-homeassistant/runs/3831046626#step:26:1`
![image](https://user-images.githubusercontent.com/5598750/136453386-363c4a34-31a5-4a10-b2ea-72dc34e9b3fa.png)

Tony763 replied:
I also added copy of logs from Mycroft, so if anything goes wrong, You can check them.

Tony763 replied:
Done

krisgesling commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,243 @@
+        HASS_SERVER: http://127.0.0.1:8123
+        HASS_TOKEN: eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJmMzI5NmEzYTA0ZDc0NTI3OGUzMDIzN2I5MDlmYmM5MSIsImlhdCI6MTU5ODczMzU5MywiZXhwIjoxOTE0MDkzNTkzfQ.HEGTdkF5tt473J3WktjUIb71x06tZwcZUjp070QzLec
+        TERM: xterm
+        MAX_LINE_LENGHT: 127
```
Little typo on LENGTH here - it's consistent throughout so would work

Tony763 replied:
Ouch :sweat_smile: Commit prepared.

Tony763 replied:
Done

krisgesling commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,243 @@
+        target_url: https://${{ github.repository_owner	}}.github.io/skill-homeassistant/${{ github.run_number }}/
+
+    #######################################################################################
+    # Section with Linters
```
Is it worth splitting this section into a separate job? This will show more clearly what failed in the Github Actions output.

Tony763 replied:
Reply below.

krisgesling commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,243 @@
+        sha: ${{github.event.pull_request.head.sha || github.sha}}
+
+    # Linters - isort
+    - name: isort
+      if: always()
+      uses: ricardochaves/python-lint@v1.4.0
+      id: isort
+      with:
+        use-pylint: false
+        use-pycodestyle: false
+        use-flake8: false
+        use-black: false
+        use-mypy: false
+        use-isort: true
+        extra-isort-options: ""
```
I haven't seen this action before - what's the benefit of running each one separately? 

Could we set pylint, flake8 and isort to true and have it run in a single step?

Tony763 replied:
I will summary this there for also other two ask as they correspond.

Each workflow or job use its own virtual environment, so If I divide them, I would have to set mycroft again as some linters also check imported modules and functions so You have two options. Disable checks or better run linters inside Mycroft's virtual enviroment.

For example `isort`: I run it separated from others and register it state as `id: isort`. With publish from above I push it to PR/Commit so you can visually see if it's run un/successful.

`isort, pycodestyle, flake8` can run on on its own, so I can run them from  `ricardochaves/python-lint@v1.4.0` role (run into own docker image). But `pylint` go into deep and output more better results, when run inside Mycroft virtual enviroment, so I have to install it on my own (An it resulted to whole lot of work as it found much more errors than when I run it in `python-lint` role.

krisgesling reviewed (commented):
First up - this is awesome!

I can't comment on each change - but really love all the code style cleanup.

I've also been thinking about setting up a Github Actions template that any Skill can grab and use - so really excited to see what you've done here. I hadn't thought about using Github Pages for the allure output either - very nice!

Tony763 commented:
Hi @krisgesling, thanks, glad you like it. I original prepared this for Travis, but on it, was impossible to upload Allure report to GitHub pages as secret was not shard to runs from PR only from commits. So useless as You want allure reports mainly from PRs. GitHub Actions is whole game changer, as no secret is needed to upload to Pages as they run in same repository as `gh-pages` branch.

Template for all skills is my intent, take this as first touch on which we all can build. Also with Mycroft-core, when new release is made, there could be workflow to crate a docker image with released version and then it can be used in skills workflows to speed up runs (skip whole cloning and installing of Mycroft core).

I started working on your comments, I made changes locally and push them in wan go.

 I will try to specify as much details as possible and add them to 'README.md' in tests, so anybody can understood them. It took me some time and many runs to figure out things so documentation is really needed.

Tony763 committed:
Switch to f-Strings

krisgesling commented:
> Also with Mycroft-core, when new release is made, there could be workflow to crate a docker image with released version and then it can be used in skills workflows to speed up runs (skip whole cloning and installing of Mycroft core).

This exists! It was getting forgotten but I've added it to our standard release process now. So anytime we release a new minor or major version of mycroft-core this docker image will get updated:
https://hub.docker.com/r/mycroftai/docker-mycroft/

Really appreciate how much time it takes to work this stuff out so really glad to hear you're documenting it as you go.

Tony763 committed:
Add regex to check ip/url

Tony763 commented:
Hi @krisgesling, check comments please, I replied to all.

Also check `README.md` inside test directory.

krisgesling commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,315 @@
+    - name: MC - copy logs
+      run: if [[ -d /var/log/mycroft ]]; then cp -r /var/log/mycroft ${{ github.workspace }}/allure-history/${{ github.run_number }}/; fi
+      shell: bash
+
```
Should this be done later in the job? Would this only copy the logs from boot, not from the test runs?

Tony763 replied:
Yeah, pasted into wrong place. :slightly_smiling_face:

krisgesling reviewed (commented):
Looking great - love the github pages deployment on your fork.

There's a couple of comments that you might have missed. Really briefly:
- Do we want to keep the unit tests? They often serve a different purpose than integration/behavioural tests.
- The `build.yml` seems to be attempting to install mycroft-core's requirements.txt a few times. Not going to hurt but probably not needed.
- May not need the sleep to wait for mycroft-core to be ready as the VK test runner won't start until Mycroft is loaded anyway.

Tony763 committed:
Add .venv and .vscode to .gitignore

krisgesling reviewed (commented):
I think that's everything addressed from what I can see - do think it's ready to merge?

Tony763 commented:
Not Yet, I missed few comments, so working on them.

Tony763 commented:
@krisgesling done. Rebased to include latest commit. Unittests fixed and added to GA workflow. Their outcome is added to allure report. All tests green.

Let me just know before merging if there is another commit added, I will rebase this a clear all linter errors. So we start with green.

Tony763 committed:
Add unittests

Return only first match

krisgesling reviewed (approved):
Looks ready to roll with a rebase :rocket:

pfefferle commented:
@Tony763 there seems to be some "conflicting files"

Tony763 committed:
Merge branch '20.08' into feature/github_actions

Tony763 commented:
Here we go, rebase did not help so I had to solve merge issues with new commit.

Workflow  OK
Linters   OK
Behave    OK
Unittests OK

Ready to go.

Tony763 commented:
Could You check @krisgesling? Another PR is on the way and I would like to make it on top of this, fixing merge issues in this beast was something this time so I would like to avoid to rebase it again :smile:

Tony763 committed:
Add TODO and contribution hint to readme

Tony763 committed:
Final clean-up

Tony763 commented:
Memo for future implementation: Always add `sudo apt update` to workflow.

krisgesling commented:
Nice little clean up of the action step names.

I've also added a branch protection rule so in future all of these tests must pass, and there must be an approving review for any PR, before they they can be merged.

Merging this now though! :rocket: :rainbow: :tada:

krisgesling commented:
Helps if I setup GH pages first...

Re-running actions now

pfefferle commented:
Hmmm, I get a lot of errors like:

`Resource not accessible by integration`

pfefferle commented:
it seems that branches/pull requests are not able to write to the `gh-pages` branch

Tony763 commented:
@pfefferle could you send PR with any small change to my repository - `feature/github-actions-pr-fix` ?

I need to test changes for fixing workflow run for PRs.

pfefferle commented:
@Tony763 done: https://github.com/Tony763/skill-homeassistant/pull/37
